tla/consistency/ExternalHistory.tla

```diff
@@ -0,0 +1,343 @@
+---- MODULE ExternalHistory ----
+\* Defines the notion of a externally observable client history and its associated properties.
+\* The history differs for traditional client histories, as clients receive responses to 
+\* transactions before they have been committed.
+
+EXTENDS Naturals, Sequences, SequencesExt, FiniteSets, FiniteSetsExt
+
+\* Event types recorded in the history
+\* Note that transaction status requests are not modelled to reduce state space
+\* Currently only read-write (Rw) transactions and read-only (Ro) transactions are modelled
+\* Both transaction types are modelled as forward-always transactions
+\* This could be extended to support more types of read-only transactions
+CONSTANTS RwTxRequest, RwTxResponse, RoTxRequest, RoTxResponse, TxStatusReceived
+
+
+\* Transaction statuses
+\* This model does not include the unknown and pending status to reduce state space
+CONSTANTS CommittedStatus, InvalidStatus
+TxStatuses == {
+    CommittedStatus,
+    InvalidStatus
+    }
+
+\* Views start at 1, 0 is used a null value
+Views == Nat
+
+\* Sequence numbers start at 1, 0 is used a null value
+SeqNums == Nat
+
+\* TxIDs consist of a view and sequence number and thus start at (1,1)
+TxIDs == Views \X SeqNums
+
+\* This models uses a dummy application where read-write transactions 
+\* append an integer to a list and then reads the list
+\* Read-only transactions simply read the list
+Txs == Nat
+
+\* History of events visible to clients
+VARIABLES history
+
+HistoryTypeOK ==
+    \A i \in DOMAIN history:
+        \/  /\ history[i].type \in {RwTxRequest, RoTxRequest}
+            /\ history[i].tx \in Txs
+        \/  /\ history[i].type \in {RwTxResponse, RoTxResponse}
+            /\ history[i].tx \in Txs
+            /\ history[i].observed \in Seq(Txs)
+            /\ history[i].tx_id \in TxIDs
+        \/  /\ history[i].type = TxStatusReceived
+            /\ history[i].tx_id \in TxIDs
+            /\ history[i].status \in TxStatuses
+
+\* History is append-only
+\* Like HistoryTypeOK, this property should always hold
+HistoryMonoProp ==
+    [][IsPrefix(history, history')]_history
+
+\* Indexes into history for events where a committed status is received
+CommittedEventIndexes == 
+    {i \in DOMAIN history: 
+        /\ history[i].type = TxStatusReceived
+        /\ history[i].status = CommittedStatus
+        }
+
+\* Transaction IDs which received committed status messages
+CommittedTxIDs ==
+    {history[i].tx_id: i \in CommittedEventIndexes}
+        
+InvalidEventIndexes == 
+    {i \in DOMAIN history: 
+        /\ history[i].type = TxStatusReceived
+        /\ history[i].status = InvalidStatus
+        }
+
+\* Transaction IDs which received invalid status messages
+InvalidTxIDs ==
+    {history[i].tx_id: i \in InvalidEventIndexes}
+
+\* Highest commit sequence number
+CommitSeqNum == 
+    IF CommittedTxIDs = {} 
+    THEN 0
+    ELSE Max({i[2]: i \in CommittedTxIDs})
```

Comment:
Style: `Max({i[2]: i \in CommittedTxIDs} \cup {0})`

Reply:
fixed in https://github.com/microsoft/CCF/pull/5699/commits/f9b1224e2c6e0d3ec8c4de299963b0e3db652de0